### PR TITLE
usaco-384

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: solutions/gold/usaco-384.mdx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	vector<vector<int>> grid(n, vector<int>(m));
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < m; j++) { read >> grid[i][j]; }
+	}
+	vector<vector<int>> is_start(n, vector<int>(m));
+	for (int i = 0; i < n; i++) {
+		for (int j = 0; j < m; j++) { read >> is_start[i][j]; }
+	}
+
+	auto to_int = [&](int r, int c) -> { return r * m + c; };
```
Comment (envyaims) [inline solutions/gold/usaco-384.mdx:68]: @SansPapyrus683 whyd you remove the `int` now it compile errors 
this aint java

Comment (SansPapyrus683) [reply solutions/gold/usaco-384.mdx:68]: oh ic
you used `auto`
sorry bout that, change it back please

**Comment by SansPapyrus683** [comment]
wait doesn't log N^2 -> 2 log N -> log N?

**Comment by envyaims** [comment]
> 

doesnt really matter, it just makes it alittle more precise

File: solutions/gold/usaco-384.mdx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	int get(int x) {
+		if (p[x] != x) { p[x] = get(p[x]); }
+		return p[x];
+	}
+
+	void unite(Edge e) {
+		int a = get(e.u);
+		int b = get(e.v);
+		if (sz[a] < sz[b]) swap(a, b);
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-384.mdx:46]: braces

File: solutions/gold/usaco-384.mdx [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+struct Edge {
+	int u, v, w;
+};
+
+struct DSU {
+	vector<int> p, sz, cnt_start;
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-384.mdx:29]: rename `cnt_start` to `unprocessed_starts` bc it seeems that's what it actually contains

File: solutions/gold/usaco-384.mdx [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	// cnt_start stores unprocessed start nodes in a component
+	DSU(int n) {
+		p.resize(n);
+		sz.resize(n, 1);
+		cnt_start.resize(n);
```
Comment (SansPapyrus683) [inline solutions/gold/usaco-384.mdx:34]:
```suggestion
	DSU(int n) : p(n), sz(n, 1), cnt_start(n) {
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
yeah just this stuff & we done

File: solutions/gold/usaco-384.mdx (hunk 3 above)
Comment (envyaims) [reply solutions/gold/usaco-384.mdx:29]: i think my comment there is good enough isnt it?

Comment (SansPapyrus683) [reply solutions/gold/usaco-384.mdx:29]: why add a comment when you could just add it into the name itself

Comment (envyaims) [reply solutions/gold/usaco-384.mdx:29]: im getting karma

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/gold/usaco-384.mdx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+		}
+	}
+
+	ll ans = 0;
+	for (Edge e : edges) {
+		dsu.unite(e);
+		int par = dsu.get(e.u);
+		if (dsu.sz[par] >= t) {
+			// add number of start nodes we havent processed * current weight
+			ans += (ll)e.w * dsu.unprocessed_starts[par];
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-384.mdx:99]:
```suggestion
			ans += (ll)(e.w * dsu.unprocessed_starts[par]);
```

**Review by ryanchou-dev — APPROVED** [review]

File: solutions/gold/usaco-384.mdx (hunk 5 above)
Comment (SansPapyrus683) [reply solutions/gold/usaco-384.mdx:99]: actually people format it like [envy does](https://en.cppreference.com/w/c/language/cast)

**Pull request merged**